storage/range_tree.go

```diff
@@ -18,31 +18,306 @@
+	}
+	node := &proto.RangeTreeNode{}
+	ok, _, err := db.GetProto(engine.RangeTreeNodeKey(*key), node)
+	if err != nil {
+		return nil, err
 	}
+	if !ok {
+		return nil, util.Errorf("Could not find the range tree node:%s",
+			engine.RangeTreeNodeKey(*key))
+	}
+	return (*rangeTreeNode)(node), nil
+}
 
+// getRootNode returns the RangeTree node for the root of the RangeTree.
+func (t *rangeTree) getRootNode(db *client.KV) (*rangeTreeNode, error) {
+	return getRangeTreeNode(db, &t.RootKey)
+}
+
+// InsertRange adds a new range to the RangeTree.  This should only be called
+// from operations that create new ranges, such as range.splitTrigger.
+func InsertRange(db *client.KV, key proto.Key) error {
+	tree, err := getRangeTree(db)
+	if err != nil {
+		return err
+	}
+	root, err := tree.getRootNode(db)
+	if err != nil {
+		return err
+	}
+	root, err = tree.insert(db, root, key)
+	if err != nil {
+		return err
+	}
+	if !root.Black {
+		// Always set the root back to black.
+		root.Black = true
+		if err = root.set(db); err != nil {
+			return err
+		}
+	}
+	if !tree.RootKey.Equal(root.Key) {
+		// If the root has changed, update the tree to point to it.
+		tree.RootKey = root.Key
+		if err = tree.set(db); err != nil {
+			return err
+		}
+	}
 	return nil
 }
+
+// insert performs the insertion of a new range into the RangeTree.  It will
+// recursivly call insert until it finds the correct location.  It will not
+// overwite an already existing key, but that case should not occur.
+func (t *rangeTree) insert(db *client.KV, node *rangeTreeNode, key proto.Key) (*rangeTreeNode, error) {
+	if node == nil {
+		// Insert the new node here.
+		node = (*rangeTreeNode)(&proto.RangeTreeNode{
+			Key: key,
+		})
+		if err := node.set(db); err != nil {
+			return nil, err
+		}
+	} else if key.Less(node.Key) {
+		// Walk down the tree to the left.
+		left, err := getRangeTreeNode(db, node.LeftKey)
+		if err != nil {
+			return nil, err
+		}
+		left, err = t.insert(db, left, key)
+		if err != nil {
+			return nil, err
+		}
+		if node.LeftKey == nil || !(*node.LeftKey).Equal(left.Key) {
+			node.LeftKey = &left.Key
+			if err := node.set(db); err != nil {
+				return nil, err
+			}
+		}
+	} else {
+		// Walk down the tree to the right.
+		right, err := getRangeTreeNode(db, node.RightKey)
+		if err != nil {
+			return nil, err
+		}
+		right, err = t.insert(db, right, key)
+		if err != nil {
+			return nil, err
+		}
+		if node.RightKey == nil || !(*node.RightKey).Equal(right.Key) {
+			node.RightKey = &right.Key
+			if err := node.set(db); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return node.walkUpRot23(db)
+}
+
+// isRed will return true only if n exists and is not set to black.
+func (n *rangeTreeNode) isRed() bool {
+	if n == nil {
+		return false
+	}
+	return !n.Black
+}
+
+// walkUpRot23 walks up and rotates the tree using the Left Leaning Red
+// Black 2-3 algorithm.
+func (n *rangeTreeNode) walkUpRot23(db *client.KV) (*rangeTreeNode, error) {
```

Comment:
Was all of this code taken near-verbatim from another trusted source or written from scratch? (I'd like to know how much scrutiny it deserves...)